条件式が Bool でない場合に rustc でエラーが出る

以下のコードをコンパイルすると
```
*a
.-$#
```
以下のエラーが出る
```
error[E0308]: mismatched types
  --> example/fizzbuzz.rs:47:14
   |
47 |           if (-v.clone()) {
   |              ^^^^^^^^^^^^ expected `bool`, found `i32`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
rustc returned with non-zero exit code
```
本来、意味解析で「条件式の型が bool になっていません」とエラーを出すべき。

--- a/src/semantics.rs
+++ b/src/semantics.rs
@@ -361,20 +361,28 @@ fn analyze_capture_condition(expr: &ast::ExprAST, capture_name: &String) -> Hash
 
     // キャプチャに型を 1 つずつ割り当てて検証
     let mut possible_types = HashSet::new();
-    Type::all_types().into_iter().for_each(|t| {
+    let mut cannot_evaluate = true;
+    for t in Type::all_types() {
         // type_validate_expr に渡す用の型ヒント
         let captures = {
             let mut captures = HashMap::new();
             captures.insert(capture_name.clone(), t);
             captures
         };
-        if let Ok(_) = type_validate_expr(expr, &captures) {
-            possible_types.insert(t);
+        if let Ok(expr_t) = type_validate_expr(expr, &captures) {
+            cannot_evaluate = false;
+            if let Type::Bool = expr_t {
+                possible_types.insert(t);
+            }
         }
-    });
+    }
 
     if possible_types.is_empty() {
-        panic!("この条件式を計算できるキャプチャ型は存在しません: {:?}", expr);
+        if cannot_evaluate {
+            panic!("この条件式を評価可能なキャプチャ型は存在しません: {:?}", expr);
+        } else {
+            panic!("結果が真偽値にならない条件式はサポートされていません: {:?}", expr);
+        }
     }
     possible_types
 }

